### ISSUE TITLE
Question: How to correctly use a single sin/cos feature for cyclical features to get better perfomance?

### ISSUE THREAD

**Issue by innjoshka** [catboost/catboost#1727]
In my data there are cyclical features for solving Regression task and I want to get maximum from it.
I have [read](https://github.com/catboost/catboost/issues/611) that @annaveronika gets perfomance from sin/cos cyclical features and @david-waterworth somehow uses it as one feature Let's take for example hour, As I understand Gradient Boosting Algorithms such as catboost [use their split rules according to one feature at a time](https://towardsdatascience.com/cyclical-features-encoding-its-about-time-ce23581845ca).  This means that they will fail to process these two features simultaneously whereas the cos/sin values are expected to be considered as one single coordinates system.

I have tried instead of categorical feature use these two features but I didn't succeed to improve results using Catboost. So, I completely believe that I need somehow to pass it as one column but I don't know how and whether it's possible 

Am I right that cyclical encoding of date features by two columns only get boosting for NN algorithms? Is it possible to combine sin/cos features into one single column and if yes, then how to do it properly? 

Thanks for any help and suggestions

**Comment by david-waterworth** [comment]
@innjoshka you're correct that each split is performed a single feature at a time, so sin/cos or categorical encoding are similar - it takes two splits for the algorithm to create a range (ie 9am-5pm) in each case. I've generally found the results to be pretty good though, sometimes one is better than another but you just have to try both (using a dev set to properly evaluate).

One option (that I've not actually evaluated as I had time of week in 15 minute samples) is to create a series of feature maps for say hour-of-day each being offset by one from the previous, i.e.

> 00:00 -> 0.0, 01:00->1.0 ... 23:00->23
> 00:00 -> 1.0, 01:00->2.0 ... 23:00->0

> 00:00 -> 23.0, 01:00->0.0 ... 23:00->22.0

This will then allow the algorithm to split on any consecutive range of hours in a single split - first by selecting one the 24 features to split on and second to split it on the start hour.

Again whether this produces a better model is dataset (and probably hyper-parameter) dependent. You may find it gives better training loss but worse test loss!

**Comment by innjoshka** [comment]
@david-waterworth Thanks for the answer.

I clarify,
sin/cos or categorical encoding are similar 

let's take example, HOUR_OF_DAY==5 hour 
>>> np.sin(5)
-0.9589242746631385
>>> np.cos(5)
0.28366218546322625

CAT_SIN/COS_HOUR = categorical_feature('[-0.9589242746631385, 0.28366218546322625]')  
CAT_HOUR = categorical_feature(5)

CAT_SIN/COS_HOUR==CAT_HOUR -- here we will get the same results

If I do, it by 2 features
NUM_SIN_HOUR =  numerical_feature(-0.9589242746631385)
NUM_COS_HOUR = numerical_feature(0.28366218546322625)
Then catboost won't be able to find the link between NUM_SIN_HOUR & NUM_COS_HOUR together because catboost performs each split is a single feature at a time, and those two features has to be interpreted together otherwise  sinus or cosinus 
separately can have exact values during the day and consequently will not have right interpretation of our cyclical feature

Sorry if I'm mistaken, this is how I understand it.

Have you used during training your model [SIN_HOUR, COS_HOUR] as two features or CAT_SIN/COS_HOUR?

One option (that I've not actually evaluated as I had time of week in 15 minute samples) is to create a series of feature maps for say hour-of-day each being offset by one from the previous, i.e.
--> I'm not sure that I understand it correctly, can ask you to create some series sample in pandas, so I can figure it out rightly? 

Thanks for your help

**Comment by david-waterworth** [comment]
First your sin and cos features should be encoded based on the period, i.e. if you have hours 0-23 then the encoded features should be

```
NUM_SIN_HOUR=np.sin(2. * np.pi * HOUR_OF_DAY/24.)
NUM_COS_HOUR=np.cos(2. * np.pi * HOUR_OF_DAY/24.)
```

Now if your model only has 2 features, you can plot them and you'll get a circle, and you can represent the splits of the tree as horizontal and vertical lines. So if we first split on the x-axis, then each of the two children is split on the y-axis you get something that looks like the image below. The black lines are the x/y axis and the circle is the encoding of time of day (it's actually a series of discrete points). The red represent the splits - so the result is 4 regions of time, which is the downside of this method, a single split can only produce a range that's symmetrical, i.e. 10am - 2pm, two splits can generate an arbitrary region for say 9am-5pm but will also produce 3 other regions.

![Screenshot from 2021-06-03 11-17-35](https://user-images.githubusercontent.com/5028974/120571442-793d5f80-c45d-11eb-9e2f-19a7c05d3594.png)

**Comment by kizill** [comment]
@david-waterworth Thank you for your help, closing as solved :)
